### PR TITLE
14-implement-the-bpref-teacher

### LINKED ISSUE #14: Implement the B-Pref teacher
### The problem
Give better human like synthetic feedback
### The solution / the feature
Implement B-Pref
Paper: [B-Pref: Benchmarking Preference-Based Reinforcement Learning](https://arxiv.org/abs/2111.03026#)
### Which Pros/Cons does this solution have?

- Better human like feedback
- Parameters to simulate different type of feedback

### PR DESCRIPTION
**Pull request by Wizzardexe**
**Issue:** Fixes #14

**What?**
- Implement the advanced Simteacher from B-Pref
- Paper: https://arxiv.org/pdf/2111.03026
- Fix an issue with the reward loss function

**How?**
- Changed the output of the preference_loss() to only the probability of the first_trajectory
- Changed the loss function to BCE, therefore we only need one probability and preference

**Ready to merge to main?**
- [ ] Linter passes
- [ ] PR Approved

### CODE REVIEW

File: framework/reward_net.py [hunk 1]
```diff
@@ -27,21 +27,19 @@ def preference_prob(self, r1, r2):
...
-        return prop1, prop2
+        return prop1
 
     def preference_loss(self, predictions, preferences):
-        # Compute loss based on human feedback
-        # predictions shape: (batch_size, 2)
-        # preferences shape: (batch_size, 2)
-        return -torch.sum(preferences * torch.log(predictions))
+        # Compute binary cross entropy loss based on human feedback
+        return -torch.mean(preferences * torch.log(predictions) + (1 - preferences) * torch.log(1 - predictions))
```
Comment (bkuen) [inline framework/reward_net.py:34]: Why do we need the mean here?

File: framework/sac_rlhf.py [hunk 2]
```diff
@@ -241,8 +254,21 @@ def get_action(self, x):
 
                 print("step", global_step, i)
 
+                SimTeacher = Teacher(
```
Comment (bkuen) [inline framework/sac_rlhf.py:257]: Please rename to sim_teacher  as this describes a concrete instance of a teacher

File: framework/sac_rlhf.py [hunk 3]
```diff
@@ -78,6 +78,19 @@ class Args:
     teacher_learning_rate: float = 1e-3
     """the learning rate of the teacher"""
 
+    # Simulated Teacher
+    teacher_sim_beta: float = -1
```
Comment (bkuen) [inline framework/sac_rlhf.py:82]: According to the paper, the teacher becomes a perfectly rational and deterministic as β →∞, whereas β = 0 produces uniformly random choices. Why did you pick -1 as a value? Is -1 even a valid value in the value range?

**Review by bkuen — COMMENTED** [review]
Looks good to me in total, just naming and one question regarding taking the mean of the cross entropy loss. Afterwards, I'll approve

File: framework/reward_net.py (hunk 1 above)
Comment (Wizzardexe) [reply framework/reward_net.py:34]: Currently, we process one preference-prediction pair at a time, so `torch.mean` is redundant. If we later extend this process to handle batches, `torch.mean` will compute the average loss across the batch.

File: framework/sac_rlhf.py (hunk 3 above)
Comment (Wizzardexe) [reply framework/sac_rlhf.py:82]: While a large positive beta also creates deterministic behavior, using beta = -1 is a convention to ensure the oracle has a standardized configuration distinct from noisy or stochastic teachers.

**Review by bkuen — APPROVED** [review]
